### ISSUE TITLE
Not working under Android Q/10?

### ISSUE THREAD

**Issue by breem42** [wolpi/prim-ftpd#181]
Hi. You app had worked very well for several years on several phones, until my Pixel 3a updated to Android 10. Then I could no longer contact the server.

I have no idea why it stopped working. I tried swiFTP and the results were the same. Maybe it's my fault. Or maybe something changed in Android 10.

I am using the Fdroid version, 5.17
The FTP client I am using is Filezilla 3.28.0 on Ubuntu 18.04.3, as patched as it can be, on an oldish Dell desktop with built in wifi. This is what I have always used. The error message Filezilla gives is:

Connection timed out after 20 seconds of inactivity
Could not connect to server

Thanks for all the work you put into this app.

**Comment by quicktrick** [comment]
The same is here. I can connect to the server, but I cannot get the file list. 

I've tried FTPDroid too, it is working, but when installing it the system warned it was build for an old Android version, and it's screen is dimmed while working.

**Comment by hedwinkoning** [comment]
The are some changes made in Android 9/10 with respect to serivces.
In ServicesStartStopUtil, at line 80

Change code line(s):

  LOGGER.debug("Use startService");
  context.startService(intent);

With

  if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
      LOGGER.debug("Use startForegroundService");
      context.startForegroundService(intent);
  } else {
      LOGGER.debug("Use startService");
      context.startService(intent);
  }

Now it should work.

**Comment by wolpi** [comment]
I cannot reproduce this in simulator. Thus I cannot verify if proposed change works. Could you post logs? That could give us hints about whats going wrong. Would you be able to install a test build uploaded here?
@hedwinkoning would you like to create a pull-request for this?

**Comment by quicktrick** [comment]
> Could you test if it solves the issue for you?

I have the same issue anyway. My FTP-client (Total Commander) connects to the server, but it "hangs" on getting the file list.

**Comment by wolpi** [comment]
the log shows basically how preferences are loaded but it does not contain hints about this error

**Comment by sumdog** [comment]
I was running into a similar issue, the server would start and I could browse everything but uploading any files always failed with a disconnect and my FTP client would retry (attempted to use ncftp, lftp and the secure clients sftp and scp).

 I tried the build @wolpi posted. That build will crash if it tries to start the secure service. Starting just the regular FTP services works, but I run into the same issue. An FTP client will start uploading a file and the connection gets dropped and nothing gets transferred (the file exists on the phone, but it's empty).

I'm on Android 9, build 53.0.A.4.114 on a Sony Xperia 10 stock/rooted via Magisk.

**Comment by wolpi** [comment]
Starts sounding like #185 and #163. Some of those issues might disapear when upgrading to current targetsdk version. That is blocked by #157.

**Comment by wolpi** [comment]
Please report back if version 6.0 solves this issue for you

**Comment by rKsanu2MMYvypWePtQWM** [comment]
Personally I can't get a directory listing with sftp, and with regular ftp it just returns empty folders. The default location selector in settings shows the same result - /storage/emulated/0 returns empty.

**Comment by hedwinkoning** [comment]
I am running latest build on an Samsung s10 with Android Q. It works fine by me. Also running it samsung secure folder and that works as well.
What mobile phone do you have? There can be differences per phone provider.

Do you use plain old file system or SAF? Also sure there is no firewall in between? You need to use wifi, mobile network does not work.

**Comment by rKsanu2MMYvypWePtQWM** [comment]
Pocophone F1 (beryllium) on Q (not MIUI, but Havoc-OS). I know this configuration is not stock at all, however other ftp solutions like built-in mixplorer ftp server works fine.
Actually, before your message I thought that SAF is only for external media (as the example states SD cards), however after clicking SAF a file selector with "allow access" showed up, after which I was able to connect successfully via SFTP without root permissions granted. So I guess something is wrong with the plain old filesystem.
And no, there is no firewall and I'm on wifi. The error shows that connection is established, but it is unable to list the directory (permission denied).

**Comment by wolpi** [comment]
Well, there is just one way to figure out what goes on on your phone: post logs here

**Comment by rKsanu2MMYvypWePtQWM** [comment]
Just tried the built-in logging. It doesn't generate the log file. What does "log to android" do? Maybe it would work better?

**Comment by wolpi** [comment]
hhmmm, as you device does not allow filesystem access it is not possible to write log file.

You can access "log to android" with developer tools included in android sdk. E.g. commandline tool adb or the IDE android-studio.

Do file management apps like amaze work on your device?

**Comment by rKsanu2MMYvypWePtQWM** [comment]
For file access I'm using MiXplorer, which only requires the Storage permission, and it works fine. Other apps that require storage access (eg. photo sharing in chat apps) work fine as well.

**Comment by wolpi** [comment]
without logs we cannot tell what goes wrong

**Comment by rKsanu2MMYvypWePtQWM** [comment]
[logcat.txt](https://github.com/wolpi/prim-ftpd/files/4272528/logcat.txt)
`192.168.0.110` is the client
Description of the events:

- 00:05 The app is started
- 00:06 The server is started (sftp mode only)
- 00:07:11 FileZilla connects, `pwd` succeds
- 00:07:13 FileZilla fails `ls` directory listing (`Error:	Unable to open .: permission denied`)
- 00:07:33 FileZilla tries again
- 00:07:53 FileZilla gives up (`Error:	Connection timed out after 20 seconds of inactivity` and `Error:	Failed to retrieve directory listing`)
- 00:08 I hit the server stop button in the app

**Comment by wolpi** [comment]
The important part is this:
`isReadable() -> false`
The system tells the app that directory would not be readable. In next days I can upload a test build here which just ignores this.

**Comment by wolpi** [comment]
Here is a testbuild:
https://c.web.de/@646396036104847907/zrsCu_mdS1K3ZtyrurvZ7A
It just ignores when system says something would not be readable with filesystem API.

**Comment by rKsanu2MMYvypWePtQWM** [comment]
It lists it, though the directory shows up as empty. Even the "0" folder doesn't exist, only "emulated".
[logcat2.txt](https://github.com/wolpi/prim-ftpd/files/4281334/logcat2.txt)

**Comment by wolpi** [comment]
Well, the system behaves as it said before: it does not allow the app to access that directory.
You have android sdk available, right?
Please open a shell on your device with `adb shell` and post the output of `ls -lha /storage/emulated/0`

**Comment by rKsanu2MMYvypWePtQWM** [comment]
![ls](https://user-images.githubusercontent.com/35463654/75919638-92c0b780-5e66-11ea-91a5-bb6dde6836b3.png)

**Comment by wolpi** [comment]
This looks normal. Other devices have a different group-name (like "everybody") but none have permission for "other".

Have a look in settings app if this app really got storage permission. Should look like this:

![grafik](https://user-images.githubusercontent.com/1812560/76029083-17c5d280-5f34-11ea-92fe-50d8e0548355.png)

**Comment by rKsanu2MMYvypWePtQWM** [comment]
Yup, I've made sure it has that permission after I noticed the bug.

**Comment by luxigo** [comment]
Does not work for me neither. Last time II just had to uninstall and reinstall.

I'm using Android 10 with kernel 4.9.118-17633310 on a Samsung S9+

Cannot see anything with version 6.1 (code: 47) in /self, /enc_emulated, /emulated and /0000-0000.

And when using SAF I have:

```
Status:	Insecure server, it does not support FTP over TLS.
Status:	Logged in
Status:	Retrieving directory listing...
Command:	PWD
Response:	257 "/" is current directory.
Command:	TYPE I
Response:	200 Command TYPE okay.
Command:	PASV
Response:	227 Entering Passive Mode (192,168,178,27,22,46)
Command:	MLSD
Response:	150 File status okay; about to open data connection.
Response:	550 Requested action not taken.
Error:	Failed to retrieve directory listing
```

**Comment by hedwinkoning** [comment]
try /sdcard as the default location to open on the mobile device

**Comment by rKsanu2MMYvypWePtQWM** [comment]
Status:	Retrieving directory listing of "/sdcard"...
Command:	cd "/sdcard"
Error:	Directory /sdcard: permission denied

**Comment by hedwinkoning** [comment]
> Status: Retrieving directory listing of "/sdcard"...
> Command: cd "/sdcard"
> Error: Directory /sdcard: permission denied

Is this still via ADB?

**Comment by rKsanu2MMYvypWePtQWM** [comment]
These are ssh terminal commands sent to the sftp server from FileZilla.

**Comment by hedwinkoning** [comment]
> These are ssh terminal commands sent to the sftp server from FileZilla.

That will not work for the same reason that in Filezilla I can not use "/" as the default remote directory. It tells me "permission denied". It does not have access to the root directory.

ADB does but ADB has near root access

**Comment by rKsanu2MMYvypWePtQWM** [comment]
what is it exactly that you want me to try? i guess i misunderstood.

**Comment by hedwinkoning** [comment]
What I want you to try is to use filezilla with the remote directory being "/sdcard"

**Comment by wolpi** [comment]
Just a blind guess: this might be related to privacy change in android 10 (google considers it bad privacy when one app is able to access all files):
https://developer.android.com/about/versions/10/privacy/changes

**Comment by hedwinkoning** [comment]
#

> Just a blind guess: this might be related to privacy change in android 10 (google considers it bad privacy when one app is able to access all files):
> https://developer.android.com/about/versions/10/privacy/changes

You guessed correct :)

**Comment by rKsanu2MMYvypWePtQWM** [comment]
> What I want you to try is to use filezilla with the remote directory set to "/sdcard"

That's exactly what I've did in that post. It gets a "permission denied".

**Comment by wolpi** [comment]
due to #196 situation should be improved in version 6.2

**Comment by wolpi** [comment]
App has been switched to use scoped files for it's own files (authorized_keys and logs). For now compatibility from #196 is still active.
